### ISSUE TITLE
LMSequenceWriter funcionality

### ISSUE THREAD

**Issue by jllombart** [Microsoft/CNTK#153]
Hi
I'm trying to make a prediction from an lm learned.
I have used the PennTreebank example to train an lstm language model.
Now i try to write the output of the PosteriorProb layer in order to have the prediction of the next word in the sequence given the previous words of the sequence.
With the default writer there is not appropriate result.
Now i'm trying to use the LMSequenceWriter (in code there is some SequenceWriter which i also try to use) but there is no output
 this is my configuration based on reader of the example

# Predict CONFIG                        

  predict = [
    action = "write"
    minibatchSize = 1
    traceLevel = 1
    epochSize = 0
    recurrentLayer = 1
    defaultHiddenActivity = 0.1
    useValidation = true
    rnnType = "CLASSLSTM"

    reader = [
        readerType = "LMSequenceReader"
        ........ # lines in example 
        ]
        sequence = [
            dim = 1
            wrecords = 2
            sectionType = "data"
        ]
        
        labelIn = [
            dim = 1
            ........ # lines in example 
        ]

        labels = [
             ........ # lines in example 
            ]
            
            category = [
                dim = 3
                sectionType = "categoryLabels"
            ]
        ]
    ]
		
    writer = [
     
        writerType = "LMSequenceWriter"
        randomize = "none"
        wordclass = "$ModelDir$/vocab.txt"

        wfile = "$OutputDir$/sequenceSentence.bin"
        wsize = 256
        wrecords = 1000
        windowSize = "$confVocabSize$"

        file = "$DataDir$/$predictFile$"

        features = [
            dim = 0
            sectionType = "data"
        ]
        
        sequence = [
            dim = 1
            wrecords = 2
            sectionType = "data"
        ]
        
        labelIn = [
            dim = 1
            labelDim = "$confVocabSize$"
            labelMappingFile = "$OutputDir$/sentenceLabels.txt"
            
            labelType = "Category"
            beginSequence = "<s>"
            endSequence = "</s>"

            elementSize = 4
            sectionType = "labels"
            
            mapping = [
                wrecords = 11
                elementSize = 10
                sectionType = "labelMapping"
            ]
            
            category = [
                dim = 11
                sectionType = "categoryLabels"
            ]	
        ]
        labels = [
            dim = 1
            labelType = "NextWord"
            beginSequence = "O"
            endSequence = "O"
            labelDim = "$confVocabSize$"

            labelMappingFile = "$OutputDir$/sentenceLabels.out.txt"

            elementSize = 4
            sectionType = "labels"
            
            mapping = [
                wrecords = 3
                elementSize = 10
                sectionType = "labelMapping"
            ]
            
            category = [
                dim = 3
                sectionType = "categoryLabels"
            ]
			
		
	  ]
		
	   PosteriorProb = [
		dim=1000
		file="$OutputDir$/outputseq.txt"
		
	   ]
		
		
      ]
]

the problem that shows is that there is no outputNodeName but it is the last PosteriorProb section

Could be this writer the correct one?
if it is not, which one is it?

If this is not the place where discuss this things just tell me and delete the post
Is there any further documentation with this kind of things?
thanks a lot

**Comment by frankseide** [comment]
Could you elaborate what problem you are encountering with the default writer?

With the default writer, you should be able to write out the sequence of posterior vectors for each input sequence, but with two caveats:

1. You currently need to make sure that it only processes one sequence at a time. For that, you need to say "nbruttsineachrecurrentiter = 1" in the reader config.

2. The LMSequenceReader has a bug that randomization cannot be disabled, which is fine for training and PPL evaluations, but not your use case.

I have fixed 1., currently in branch fseide/s2s, so that the default writer handles multiple sequences in a single minibatch, and you can define custom delimiter strings to mark where a new sequence begins), and 2. can be worked around by also setting "cacheBlockSize = 1". I hope to get this fix landed in master on Monday. I will now look into revising the RNN.cntk example to include a default writer section as well.

**Comment by frankseide** [comment]
OK, trying to add a "write" command to the RNN.cntk sample, I now understand this better.

First, you are using a class-based LM. For that, writing PosteriorProb is not correct. PosteriorProb only represents the class-conditional probabilities, (1) without class probabilities, and (2) unhelpfully renormalized as if they were posteriors. If you are not using classes, PosteriorProb is what you want, but with classes, you currently will need to do some math yourself. You need to
* write out ClassPostProb as well;
* renormalize PosteriorProb into class-conditional probabilities (divide the value for word i by the sum over all values for words belong to the same class as i); and
* multiply those with the correct class probs.
The vocab.txt file generated by the "writeWordAndClassInfo" step tells you the word and class mappings; first column is the word id, and last column the class id.

However, even if you do that, all you get is the per-word conditional distribution for a fixed history (the ground truth). It is not clear to me how one would use this other than for measuring perplexity.

I see two main ways to apply LM scores: (1) embedded inside a decoder (e.g. lattice rescoring) and (2) rescoring of flat N-best lists. (1) requires a tight integration API that allows to store state that depends on a hypothesized word history. This does not existing CNTK currently, but I will make sure this is covered in our CNTK librarification work that we are doing currently.

That leaves (2). For (2), you don't need per-word probs, but only the aggregate probability for the entire sentence.

That can be computed using the criterion node, which computed the aggregate logP over all tokens in the minibatch. This is equal to the sentence score if the minibatch consists of one sentence only. This is how to do it:
* create your N-best list as a text file, one line per hypothesis (we are going to compute one sentence score per hypothesis)
* run the "write" command with nbruttsineachrecurrentiter = 1 (=> one sentence per minibatch) and cacheBlockSize = 1 (no randomization)
* choosing the training criterion node as the output (outputNodeNames = TrainNodeClassBasedCrossEntropy)

This should generate an output file with the aggregate log sentence probabilities of your N-best hypotheses, which you can now post-process/rerank with your own code.

It am curious whether this works for you.

(One caveat/bug: The LMSequenceReader, in an ill-conceived attempt at being helpful, will skip any empty sentence, that is, sentences with 2 or less tokens. If you have empty hypotheses, this would get our output out of sync, so please look out for that.)

**Comment by jllombart** [comment]
Ok, I think that I have already understand you.

I know that I was training a language model with class clustering in training, but the thing was that I'm not interested on last layer with the class clustering layer. I was thinking that I was obtaining the previous layer where there is the hole vocabulary, in my toy example 1000 words. That was the layer I want to write. The log-likelihood of the words. Yes for the values that I saw I was thinking that I have to renormalizes in any way, thank for the tip.

The experiment I was trying to perform is whether there is given an unfinished sentence predict the next possible word  that could fit in this sentence, or a subset of words. For example if the test sentence, the given sentence is " I was playing ", I hope to obtain a subset of words that may fit next to the given sentence like " football, tennis, cards "

What I tried was to train language model like it is in PennTreebank example. Then I perform a test in with I provide to de lstm architecture the first words in the sentence and eval this sentence. I suppose that the last eval value was the word that could fit next. For that, lastly, I use UCI writer to write the penultimate layer with 1000 outputs  (the same as words in vocab, that I thought that is PosteriorProb layer) without randomize. Then take the last frame (The writer wrote one output per word, so the last frame is the one I'm interested to), and finally sort the vector to see which index have greater likelihood and look for in vocab.txt which words are those index. Those words would be the words that fit better in the sentence.  

Those cases  which you talk about are in order to obtain an log posterior that allows you to evaluate how fine fits the language model to one sentence (or better how well fits your hypothesis sentence with your model) like in automatic speech recognition. Although this explanation clarifies me a lot.

Now I'm thinking that I need other network architecture, something like sequence to item architecture. Something like give to the net an un finished sentence and the next word as target. This is what I will try next.

Thank you a lot for your attention

**Comment by dongyu888** [comment]
If your vocabulary size is just 1000, you don’t need to use the class-based LM at all. just use the normal cross entropy training node and the normal softmax node as the output.

**Comment by frankseide** [comment]
I see. You should be able to write this information with the stock output writer, no need for the UCIWriter. Although if that works for you, you can use it of course. I would use LMSequenceReader with labelType = "none" for the output labels, and stock output writer (no “writer” section) for writing the PosteriorProb. You would get one looong line of 1000 numbers per input word, and if in the “write” action you say “format = [ sequenceEpilogue = "\n\n" ]” you get an additional newline after each sentence.

You will get a collection of a class-conditional probabilities concatenated, and renormalized over all, which is not what you want. You’d need to renormalize. Or try without classes.

**Comment by wolfma61** [comment]
@jlombart - did this answer your questions?

**Comment by jllombart** [comment]
i need some work but I think this will solve my problem. When I have any result I'll say in what way worked for me

**Comment by frankseide** [comment]
Thanks, appreciated!!
